### PR TITLE
nixos/tang: create module for tang server

### PR DESCRIPTION
**Pull request by jfroche**
This commit adds a module for the tang server and the related nixos test.

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by jfroche** [comment]
Result of `nixpkgs-review pr 247037` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>

File: nixos/modules/services/security/tang.nix [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+  };
+  config = mkIf cfg.enable {
+    users = {
+      users.tang = {
+        group = "tang";
+        isSystemUser = true;
+      };
+      groups.tang = { };
+    };
```
Comment (zimbatm) [inline nixos/modules/services/security/tang.nix:31]: Is it not possible to use DynamicUser instead?

File: nixos/modules/services/security/tang.nix [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+    environment.systemPackages = [ pkgs.tang ];
+
+    systemd.services."tangd@" = {
+      description = "Tang server";
+      path = [ pkgs.jose pkgs.tang ];
+      preStart = ''
+        if ! test -n "$(${pkgs.findutils}/bin/find /var/lib/tang -maxdepth 1 -name '*.jwk' -print -quit)"; then
+          ${cfg.package}/libexec/tangd-keygen /var/lib/tang
```
Comment (zimbatm) [inline nixos/modules/services/security/tang.nix:40]:
```suggestion
        if ! test -n "$(${pkgs.findutils}/bin/find "$STATE_DIRECTORY" -maxdepth 1 -name '*.jwk' -print -quit)"; then
          ${cfg.package}/libexec/tangd-keygen "$STATE_DIRECTORY"
```

File: nixos/modules/services/security/tang.nix [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+      '';
+      serviceConfig = {
+        StandardInput = "socket";
+        StandardOutput = "socket";
+        StandardError = "journal";
+        User = "tang";
+        Group = "tang";
+        StateDirectory = "tang";
+        StateDirectoryMode = "700";
+        ExecStart = "${cfg.package}/libexec/tangd /var/lib/tang";
```
Comment (zimbatm) [inline nixos/modules/services/security/tang.nix:51]:
```suggestion
        ExecStart = "${cfg.package}/libexec/tangd %S/tang";
```

File: nixos/modules/services/security/tang.nix [hunk 4]
```diff
@@ -0,0 +1,66 @@
+{ config, lib, pkgs, ... }:
+with lib;
+let
+  cfg = config.services.tang;
+in
+{
```
Comment (zimbatm) [inline nixos/modules/services/security/tang.nix:6]: add yourself to the meta.maintainers?

**Review by zimbatm — COMMENTED** [review]
nice

**Comment by kravemir** [comment]
Thank you for this work! 

I've copy-pasted the module code to config for installation of my new network-presence (low performance) machine in https://github.com/kravemir/nixos-config/commit/7b5a0fe9a8016fba355053d3b1cfa75305e92628, and it works successfully.

Is there any expectation when this PR will be merged into nixpkgs?

File: nixos/modules/services/security/tang.nix (hunk 1 above)
Comment (jfroche) [reply nixos/modules/services/security/tang.nix:31]: Yep good call, done

File: nixos/modules/services/security/tang.nix (hunk 2 above)
Comment (jfroche) [reply nixos/modules/services/security/tang.nix]: Tang creates the key on its own anyway

**Comment by jfroche** [comment]
Note that #257525 goes further in the configuration of nixos to use clevis.
This PR sets the bases to deploy the tang service in a secure way. Maybe we should merge both work ?

File: nixos/modules/services/security/tang.nix [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+      type = types.listOf types.str;
+      description = ''
+        Whitelist a list of address prefixes.
+        Preferably, internal addresses should be used.
+      '';
+    };
+
+  };
+  config = mkIf cfg.enable {
+    environment.systemPackages = [ pkgs.tang ];
```
Comment (ElvishJerricco) [inline nixos/modules/services/security/tang.nix:35]: Should be `cfg.package`, no?

File: nixos/modules/services/security/tang.nix [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+        ProtectKernelTunables = true;
+        ProtectProc = "invisible";
+        ProtectSystem = "strict";
+        RestrictNamespaces = true;
+        RestrictRealtime = true;
+        RestrictSUIDSGID = true;
+        SystemCallArchitectures = "native";
+        SystemCallFilter = [ "@system-service" "~@privileged" "~@resources" ];
+      };
+    };
```
Comment (ElvishJerricco) [inline nixos/modules/services/security/tang.nix:78]: It's good that you've added all these protections on the service. I still wonder if we should use the included unit files from the tang package and add these extra parts as overrides. Though, I see you're using `$STATE_DIRECTORY`, which is a much better chioce than the hard coded `/var/db/tang` that the included unit uses.

File: nixos/modules/services/security/tang.nix [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+        StandardError = "journal";
+        User = "tang";
+        Group = "tang";
+        DynamicUser = true;
+        StateDirectory = "tang";
+        RuntimeDirectory = "tang";
+        StateDirectoryMode = "700";
+        UMask = "0077";
+        CapabilityBoundingSet = [ "" ];
+        ExecStart = "${cfg.package}/libexec/tangd $STATE_DIRECTORY";
```
Comment (ElvishJerricco) [inline nixos/modules/services/security/tang.nix:52]: I think the syntax for using an environment variable in `ExecStart` is `${STATE_DIRECTORY}`, not `$STATE_DIRECTORY`, isn't it? Regardless, `%S` always expands to the state directory, so it's probably best to use that.

File: nixos/modules/services/security/tang.nix (hunk 5 above)
Comment (jfroche) [reply nixos/modules/services/security/tang.nix]: Good catch, thank you

File: nixos/modules/services/security/tang.nix (hunk 6 above)
Comment (jfroche) [reply nixos/modules/services/security/tang.nix]: Yes, the definition available in their repository is really basic.

File: nixos/modules/services/security/tang.nix (hunk 7 above)
Comment (jfroche) [reply nixos/modules/services/security/tang.nix]: :+1:

**Review by zimbatm — APPROVED** [review]
LGTM!

File: nixos/modules/services/security/tang.nix [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+
+    systemd.services."tangd@" = {
+      description = "Tang server";
+      path = [ cfg.package ];
+      serviceConfig = {
+        StandardInput = "socket";
+        StandardOutput = "socket";
+        StandardError = "journal";
+        User = "tang";
+        Group = "tang";
```
Comment (JulienMalka) [inline nixos/modules/services/security/tang.nix:48]: Do we need this if we have `DynamicUser=true` ?

File: nixos/modules/services/security/tang.nix [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+      wantedBy = [ "sockets.target" ];
+      socketConfig = {
+        ListenStream = cfg.listenStream;
+        Accept = "yes";
+        IPAddressDeny = "any";
+        IPAddressAllow = cfg.ipAddressAllow;
+      };
+    };
+  };
+  meta.maintainers = with lib.maintainers; [ jfroche ];
```
Comment (JulienMalka) [inline nixos/modules/services/security/tang.nix:96]:
```suggestion
  meta.maintainers = with lib.maintainers; [ jfroche julienmalka ];
```
If that's alright to you I'd be glad to be added to the maintainers as well.

**Review by JulienMalka — COMMENTED** [review]
Would be good to add the test to `passthru` in the `tang` package

**Comment by JulienMalka** [comment]
@ofborg test

**Comment by JulienMalka** [comment]
@ofborg test tang

File: nixos/modules/services/security/tang.nix (hunk 8 above)
Comment (jfroche) [reply nixos/modules/services/security/tang.nix]: Not really. I was thinking that maybe we should give to another user/group the permission to rotate the keys but we would have to create the static user/group.

**Review by JulienMalka — APPROVED** [review]
This looks good to me, thanks you!

**Pull request merged**